dtm0/dtx.c

```diff
@@ -0,0 +1,530 @@
+/* -*- C -*- */
+/*
+ * Copyright (c) 2013-2020 Seagate Technology LLC and/or its Affiliates
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * For any questions about this software or licensing,
+ * please email opensource@seagate.com or cortx-questions@seagate.com.
+ *
+ */
+
+
+
+/**
+ * @addtogroup dtm
+ *
+ * @{
+ */
+
+#define M0_TRACE_SUBSYSTEM M0_TRACE_SUBSYS_DTM0
+#include "dtm0/dtx.h"
+#include "lib/assert.h"   /* M0_PRE */
+#include "lib/memory.h"   /* M0_ALLOC */
+#include "lib/errno.h"    /* ENOMEM */
+#include "lib/trace.h"    /* M0_ERR */
+#include "be/dtm0_log.h"  /* dtm0_log API */
+#include "conf/helpers.h" /* proc2srv */
+#include "reqh/reqh.h"    /* reqh2confc */
+#include "rpc/rpc_machine.h" /* rpc_machine */
+#include "dtm0/fop.h"     /* dtm0 req fop */
+#include "dtm0/service.h" /* m0_dtm0_service */
+
+static struct m0_sm_state_descr dtx_states[] = {
+	[M0_DDS_INIT] = {
+		.sd_flags     = M0_SDF_INITIAL | M0_SDF_FINAL,
+		.sd_name      = "init",
+		.sd_allowed   = M0_BITS(M0_DDS_INPROGRESS),
+	},
+	[M0_DDS_INPROGRESS] = {
+		.sd_name      = "inprogress",
+		.sd_allowed   = M0_BITS(M0_DDS_EXECUTED, M0_DDS_FAILED),
+	},
+	[M0_DDS_EXECUTED] = {
+		.sd_name      = "executed",
+		.sd_allowed   = M0_BITS(M0_DDS_EXECUTED_ALL),
+	},
+	[M0_DDS_EXECUTED_ALL] = {
+		.sd_name      = "executed-all",
+		.sd_allowed   = M0_BITS(M0_DDS_STABLE),
+	},
+	[M0_DDS_STABLE] = {
+		.sd_name      = "stable",
+		.sd_allowed   = M0_BITS(M0_DDS_DONE),
+	},
+	[M0_DDS_DONE] = {
+		.sd_name      = "done",
+		.sd_flags     = M0_SDF_TERMINAL,
+	},
+	[M0_DDS_FAILED] = {
+		.sd_name      = "failed",
+		.sd_flags     = M0_SDF_TERMINAL | M0_SDF_FAILURE
+	}
+};
+
+static struct m0_sm_trans_descr dtx_trans[] = {
+	{ "populated",  M0_DDS_INIT,         M0_DDS_INPROGRESS   },
+	{ "executed",   M0_DDS_INPROGRESS,   M0_DDS_EXECUTED     },
+	{ "exec-all",   M0_DDS_EXECUTED,     M0_DDS_EXECUTED_ALL },
+	{ "exec-fail",  M0_DDS_INPROGRESS,   M0_DDS_FAILED       },
+	{ "stable",     M0_DDS_EXECUTED_ALL, M0_DDS_STABLE       },
+	{ "prune",      M0_DDS_STABLE,       M0_DDS_DONE         }
+};
+
+struct m0_sm_conf m0_dtx_sm_conf = {
+	.scf_name      = "dtm0dtx",
+	.scf_nr_states = ARRAY_SIZE(dtx_states),
+	.scf_state     = dtx_states,
+	.scf_trans_nr  = ARRAY_SIZE(dtx_trans),
+	.scf_trans     = dtx_trans,
+};
+
+M0_INTERNAL void m0_dtm0_dtx_domain_init(void)
+{
+	if (!m0_sm_conf_is_initialized(&m0_dtx_sm_conf))
```

Comment:
m0_sm_conf_is_initialized() is used to ASSERT that conf if already initialized.
We usually don't use it in a condition.

If it is already m0_sm_conf_is_initialized(), a second call is a BUG.

Reply:
A second call is possible in UTs as client and server share this conf in a single process. We have a helper dtm_in_ut() but adding it there makes condition unclear, I think it's better to leave as is now.